client/src/commands/getTcb.ts

```diff
@@ -0,0 +1,120 @@
+/**
+ * @license
+ * Copyright Google Inc. All Rights Reserved.
+ *
+ * Use of this source code is governed by an MIT-style license that can be
+ * found in the LICENSE file at https://angular.io/license
+ */
+
+import * as vscode from 'vscode';
+
+import * as lspext from '../../common/lsp_extensions';
+import {AngularLanguageClient} from '../client';
+
+const TCB_HIGHLIGHT_DECORATION = vscode.window.createTextEditorDecorationType({
+  backgroundColor: new vscode.ThemeColor('editor.selectionHighlightBackground'),
+});
+
+class SourceMap {
+  private readonly lines: Array<readonly[/* line offset */ number, /* line length */ number]>;
+  constructor(source: string) {
+    let offset = 0;
+    this.lines = source.split('\n').map(line => {
+      const lineLength = line.length + 1 /* ='\n' */;
+      const lineOffsetAndLength = [offset, lineLength] as const;
+      offset += lineLength;
+      return lineOffsetAndLength;
+    });
+  }
+
+  /**
+   * Converts a character offset in the source text to a `Position` consisting
+   * of a line and column.
+   */
+  toPosition(offset: number): vscode.Position {
+    const lineNumber = this.lines.findIndex(
+        ([lineOffset, lineLen]) => offset >= lineOffset && offset < lineOffset + lineLen);
+    if (lineNumber === -1) throw new Error(`${offset} is out of bounds`);
+    const lineOffset = this.lines[lineNumber][0];
+    return new vscode.Position(lineNumber, offset - lineOffset);
+  }
+}
+
+/**
+ * Retrieves and displays the TCB corresponding to the template (if any) under the active
+ * editor cursor (if any).
+ * If a TCB is not found or cannot be displayed, this operation exits silently.
+ * @param ngClient LSP client for the active session
+ * @param didRequestTcb persistent event from which notifications of
+ *        requested/udpated TCB virtual documents will be fired, prompting
+ *        VSCode to update the content of the document.
+ *        Subsequent calls of `executeGetTemplateTcb` should use the same value
+ *        for this parameter.
+ */
+async function executeGetTemplateTcb(
+    ngClient: AngularLanguageClient,
+    didRequestTcb: vscode.EventEmitter<vscode.Uri>): Promise<void> {
+  const activeEditor = vscode.window.activeTextEditor;
+  const client = ngClient.languageClient;
+  if (activeEditor === undefined || client === null) {
+    return;
+  }
+
+  const tcb = await client.sendRequest(lspext.GetTcb, {
+    textDocument: client.code2ProtocolConverter.asTextDocumentIdentifier(activeEditor.document),
+    position: activeEditor.selection.active,
+  });
+  if (tcb === undefined) {
+    return;
+  }
+
+  // Allocate a provider of documents corresponding to the `ng` URI scheme,
+  // which we will use to provide a virtual document with the TCB contents.
+  const tcbProvider: vscode.TextDocumentContentProvider = {
```

Comment:
That is one reason; another was to support syntax highlighting IIRC (but maybe Im wrong). I will follow up and add a comment.

Reply:
Nevermind I think I did this just because of the readonly mode as you mentioned, not sure why I was thinking of highlighting. added a comment a link to the issue.